lib/collection/src/shards/replica_set/mod.rs

```diff
@@ -655,9 +662,28 @@ impl ShardReplicaSet {
 
     /// Check if the are any locally disabled peers
     /// And if so, report them to the consensus
-    pub async fn sync_local_state(&self) -> CollectionResult<()> {
-        for failed_peer in self.locally_disabled_peers.read().iter() {
-            self.notify_peer_failure(*failed_peer);
+    pub async fn sync_local_state<F>(&self, get_shard_transfer: F) -> CollectionResult<()>
+    where
+        F: Fn(ShardTransferKey) -> Option<ShardTransfer>,
+    {
+        for &failed_peer in self.locally_disabled_peers.read().iter() {
+            self.notify_peer_failure(failed_peer);
+
+            let key = ShardTransferKey {
+                shard_id: self.shard_id,
+                from: failed_peer,
+                to: self.this_peer_id(),
```

Comment:
I am not sure it should be `this_peer_id`. Failure can be reported by the third node, which may not have any shards at all. I think we should cancel all transfers from `failed_peer` no matter destination.

Reply:
But in that case, we would be cancelling shard transfers of _other_ shards from `failed_peer`, right? Cause there can't be multiple shard transfers of `self.shard_id` from `failed_peer`. (Also, not sure if there can be two different shard transfers from `failed_peer` to `this_peer_id`. If so, then this is even worse.)

Seems safer to only cancel `self.shard_id`/`this_peer_id` transfer. Like, if `failed_peer` _is_ available, but under high load, and it only fail _sometimes_, cancelling (and then restarting) all shard transfers will probably make everything even worse. If we only cancel `self.shard_id`/`this_peer_id`, then there is a chance that at least some of the transfers may finish successfully.

Reply:
>  And if all shard transfers fail, then each shard would cancel its own transfer.

Local state doesn't exist on all peers, the situation possible, where there are 3 nodes: A, B, C

The transfer A->B happens and A shuts down. Replica only exist on A and B, C is empty.
But the update can still come to C, so local state will exist on C.
In this situation we still want to cancel transfer from A->B.
No other node won't be able to cancel transfer, cause no one else knows that the inconsistency exist.

Reply:
Another thing:
If we would cancel all transfers from a node, wouldn't it be possible for a badly behaving node to constantly break _all_ ongoing transfers while they would otherwise succeed.

Badly behaving node as in: flaky network connection or something similar.

I'd expect that only cancelling A->B transfers would be a safer bet then.

Reply:
Done.